src/services/signatureHelp.ts

```diff
@@ -375,23 +375,59 @@ namespace ts.SignatureHelp {
                 const typeParameters = (candidateSignature.target || candidateSignature).typeParameters;
                 signatureHelpParameters = typeParameters && typeParameters.length > 0 ? map(typeParameters, createSignatureHelpParameterForTypeParameter) : emptyArray;
                 suffixDisplayParts.push(punctuationPart(SyntaxKind.GreaterThanToken));
-                const parameterParts = mapToDisplayParts(writer =>
-                    typeChecker.getSymbolDisplayBuilder().buildDisplayForParametersAndDelimiters(candidateSignature.thisParameter, candidateSignature.parameters, writer, invocation));
+                const parameterParts = mapToDisplayParts(writer => {
+                    writer.writePunctuation("(");
+                    let hadPrevious = false;
+                    if (candidateSignature.thisParameter) {
+                        typeChecker.parameterToString(candidateSignature.thisParameter, invocation, NodeBuilderFlags.OmitParameterModifiers | NodeBuilderFlags.IgnoreErrors, writer);
+                        hadPrevious = true;
+                    }
+                    for (const param of candidateSignature.parameters) {
+                        if (hadPrevious) {
+                            writer.writePunctuation(",");
+                            writer.writeSpace(" ");
+                        }
+                        typeChecker.parameterToString(param, invocation, NodeBuilderFlags.OmitParameterModifiers | NodeBuilderFlags.IgnoreErrors, writer);
+                        hadPrevious = true;
+                    }
+                    writer.writePunctuation(")");
+                });
                 addRange(suffixDisplayParts, parameterParts);
             }
             else {
                 isVariadic = candidateSignature.hasRestParameter;
-                const typeParameterParts = mapToDisplayParts(writer =>
-                    typeChecker.getSymbolDisplayBuilder().buildDisplayForTypeParametersAndDelimiters(candidateSignature.typeParameters, writer, invocation));
+                const typeParameterParts = mapToDisplayParts(writer => {
+                    if (candidateSignature.typeParameters && candidateSignature.typeParameters.length) {
+                        writer.writePunctuation("<");
+                        let hadPrevious = false;
+                        for (const param of candidateSignature.typeParameters) {
+                            if (hadPrevious) {

```

Comment:
it's a tiny bit concerning that the dissolution of the symbol display builder caused the dispersal of very specific display code over the code base. Specifically, I saw three new copies of this parameter-display loop. Is there any way to consolidate this somewhere else?

Reply:
Out emitter cannot emit a list directly, as a list is not a node kind... so these exist mostly to work around how I can't pass a `NodeArray` and list format into the emitter directly... Would it make sense to expose a `writeList`/`printList` alongside `writeNode`/`printNode`? Thinking on it, it probably does.